### ISSUE TITLE
Load arrow tables into Perspective without converting to a binary buffer

### ISSUE THREAD

**Issue by sc1f** [finos/perspective#1157]
### Description of Problem:

Perspective expects arrow data to be loaded as an ArrayBuffer in Javascript, and a binary string in Python. This requires, in PyArrow at least, a few lines to convert an Arrow `Table` into binary:

```python
stream = pa.BufferOutputStream()
writer = pa.RecordBatchStreamWriter(stream, arrow_table.schema)
writer.write_table(arrow_table)
writer.close()
perspective_table = perspective.Table(stream.getvalue().to_pybytes())
```

When loading arrow Tables, I expect Perspective to be compatible with Arrow Tables without having to do any conversion. The requirement of `bytes` for an Arrow binary is outlined slightly in the Python user guide, but the conversion process from an Arrow Table -> bytes is unclear.

### Potential Solutions:

Write in an Arrow Table to binary conversion layer solely in the binding layer (using PyArrow or Arrow typescript), which would be simple but incomplete (would require reimplementation in future binding languages) and less performant. If there is a way to transfer a pointer to an arrow table from the binding layer into C++, then we should be able to write the conversion entirely in C++. We already malloc and memcpy the arrow binary from JS/Python into C++, so there might be something already there worth looking into.

**Comment by timkpaine** [comment]
There is a way to transfer the pointer, I discussed this with @wesm a while back and it was eventually implemented to optimize transferring arrow tables between R and python. Let me look up the details of that thread

**Comment by timkpaine** [comment]
Looking at the code it looks simple to add, I will look into it this week and do a PR if I get something useable.

Tim Paine
tim.paine.nyc
908-721-1185

**Comment by timkpaine** [comment]
`unwrap_array` was the original way, but we weren't interested at the time because it required linking against `libarrow_python`. But since we do that anyway it should be ok
https://arrow.apache.org/docs/python/extending.html#_CPPv412unwrap_arrayP8PyObjectPNSt10shared_ptrI5ArrayEE

**Comment by sc1f** [comment]
@timkpaine is there a way to do the same in the WASM? I'm personally ok with only being able to load pyarrow Tables in our python binding but for the sake of completeness it would be helpful in WASM

**Comment by wesm** [comment]
You should be able to use the C interface now. cc @pitrou @nealrichardson

**Comment by timkpaine** [comment]
Non-c api but still doesn't require linking or knowing about pyarrow from the c++ side which was our goal:
https://github.com/timkpaine/arrow-cpp-python-nocopy

**Comment by timkpaine** [comment]
I don't think this is stable, I'm going to implement raw-pointer conversion. Does this sound right @wesm @pitrou @neilrichardson e.g. for pyarrow -> c++ across C ABI


```
# public-api.pxi
cdef api uintptr_t pyarrow_unwrap_table_rawpointer(object table):
    cdef shared_ptr[CTable] tab
    tab = pyarrow_unwrap_table(table)
    return <uintptr_t>tab.get()


# table.pxi

    @property
    def __pointer__(self):
        """
        Raw pointer of the underlying arrow::Table, for use via the C API.

        Returns
        -------
        int

        Examples
        --------
        >>> import pyarrow as pa
        >>> import pandas as pd
        >>> df = pd.DataFrame({'n_legs': [2, 4, 5, 100],
        ...                    'animals': ["Flamingo", "Horse", "Brittle stars", "Centipede"]})
        >>> table = pa.Table.from_pandas(df)
        >>> table.__pointer__
        """
        return pyarrow_unwrap_table_rawpointer(self)
```

(other direction, raw pointer from C++ to pyarrow is slightly harder)

**Comment by timkpaine** [comment]
oh wait maybe its just https://github.com/apache/arrow-nanoarrow 😆

**Comment by timkpaine** [comment]
Ok I think this is an example now:
https://github.com/timkpaine/arrow-cpp-python-nocopy

It does both parts in one, but basically the README explains it:

```
This utilizes Apache Arrow's ABI-stable C interface, which would, for example, allow your combined C++/python extension to compile and link against one version of arrow and be utilized with pyarrow which would otherwise be ABI-incompatible.
```

Does this seem like the right thing to be doing @wesm @pitrou @nealrichardson? AFAIK you need to force your python extension to interact with your C++ library via the arrow C ABI, otherwise your extension might pickup a `libarrow` from `pyarrow` that is C++ ABI incompatible with the `libarrow` your C++ code built against?

**Comment by davenza** [comment]
@timkpaine (@wesm @pitrou too) I checked your solution in [https://github.com/timkpaine/arrow-cpp-python-nocopy](https://github.com/timkpaine/arrow-cpp-python-nocopy), but it is a bit confusing for me and I believe there is something wrong. I would be grateful if you could give me some advice.

I have Arrow C++ 13 installed on my system. I can compile your example with Arrow C++ 13, and then I install it along with pyarrow 13 in a virtualenv. This works fine.

Then, I force the installation of pyarrow 12 in the virtualenv (keeping your example compiled with Arrow C++ 13). In this case, there is a `MemoryError: std::bad_alloc` in `test_create_schema_in_python` and a segmentation fault in `test_create_schema_in_cpp`.

I believe this is what happens:
- When `arrow_python_nocopy` is imported in Python, then libarrow.so.1300 is loaded (the library is compiled linking to that version). 
- Then, when pyarrow is imported (`import pyarrow as pa`), libarrow.so.1200 is imported (from the site-packages/pyarrow).
- Now, we have two different versions of libarrow.so loaded and this leads to a segmentation fault.

I may be misunderstanding something, but how does your example solve this problem using the C ABI? 

In fact, you keep using `unwrap_array()` from libarrow_python to do the  C++ <-> Python conversions (I don't see the C ABI being used in the pybind11 caster).

**Comment by timkpaine** [comment]
> When arrow_python_nocopy is imported in Python, then libarrow.so.1300 is loaded (the library is compiled linking to that version).

The user library should pull it from pyarrow instead of the system. That repo is me testing stuff, it might have more or less content than is necessary to do this properly.

**Comment by davenza** [comment]
Ok, so the library will always be compiled with the same version as pyarrow.

I probably misunderstood it, because I thought that the C ABI could be used to make different versions of Arrow work together (one version for the library and other version for pyarrow).

Anyway, thank you @timkpaine .

**Comment by pitrou** [comment]
Le 07/10/2023 à 11:21, David Atienza a écrit :
> Ok, so the library will always be compiled with the same version as pyarrow.
> 
> I probably misunderstood it, because I thought that the C ABI could be 
> used to make different versions of Arrow work together (one version for 
> the library and other version for pyarrow).

It can, but you have to be careful with library loading issues. In 
particular, loading two different versions *from Python* is probably not 
possible, because the symbols will clash.

**Comment by davenza** [comment]
For future reference for other readers who have stumbled upon the same problem, I have been thinking about how we could design an Arrow C++ <-> Python library that is decoupled from the installed pyarrow version, and now I think it is not possible.

Even if we solve the name clashes caused by the different versions of libarrow.so (mangling the symbols with the version, for example), the big problem here is that the [ABI is not stable](https://www.mail-archive.com/dev@arrow.apache.org/msg24282.html), so the library should be recompiled when the pyarrow version changes. 

For example, I could compile a library that connects Arrow version X to pyarrow version Y using the C ABI. However, this compiled library would not work if I change the pyarrow version to Z (until I recompile and link to the Z version). This makes really really hard to decouple the library from the pyarrow version.

**Comment by pitrou** [comment]
> the big problem here is that the [ABI is not stable](https://www.mail-archive.com/dev@arrow.apache.org/msg24282.html),

There is a misunderstanding here. The _Arrow C++ ABI_ is not stable. The _Arrow C Data Interface_, however, is a stable ABI by design. This is even spelled out in the spec for it:

> Goals
> * Expose an ABI-stable interface.

https://arrow.apache.org/docs/format/CDataInterface.html#goals

**Comment by davenza** [comment]
Yes, I know that the C ABI is stable, but you need the C++ and Cython ABIs to pass the data between C++ and Python.

As I see it, this is what you need to transform a C++ Arrow array of version X to a pyarrow.Array of version Y (back and forth):

```
              std::shared<arrow::Array>               (version X, C++ array, libarrow.so.X)
                           ^
                           |
    ImportArray()          |          ExportArray()   (you need this functions from libarrow.so.X)
                           |
                           v
              struct ArrowArray*          (C ABI, this should be compatible in libarrow.so.X and libarrow.so.Y)
                           ^
                           |
    ExportArray()          |          ImportArray()   (you need this functions from libarrow.so.Y)
                           |
                           v
              std::shared<arrow::Array>               (version Y, pyarrow version, libarrow.so.Y)
                           ^
                           |
    unwrap_array()         |          wrap_array()    (you need this functions from libarrow_python.so.Y)
                           |
                           v
              PyObject*                               (version Y, this is a pyarrow.Array in Python)

```
I hope the diagram is understandable. The functions on the left, transform the data from Python to C++, and the functions on the right do the opposite. As you can see, you need to link two versions of libarrow.so and one version of libarrow_python.so.

If the C++ and Cython ABIs were stable, I could statically link the Arrow C++ version X (using some type of mangling for the version), and then I could use my library with any pyarrow version. Since the C++ and Cython ABIs are not stable, I have to recompile each time I change the pyarrow version to link to the new version.

Of course, I understand the difficulties of creating a stable ABI for C++ and Cython, so I think this is not very realistic (especially in the short-term). However, I am concerned about how Arrow can create a rich ecosystem if two libraries that depend on two different versions of pyarrow cannot coexist in the same virtualenv.

Is there anything obvious that I am not seeing @pitrou? At this point I do not see a winning solution for the libraries that need C++ <-> Python interoperability in the same process.

**Comment by timkpaine** [comment]
your c++ library statically links libarrow, you ship a separate python library which dynamically links against libarrow and libarrow_python but does not vendor them. Pyarrow loads its libarrow and libarrow_python, then your python library uses those to create the C api data structures which it hands to your c++ library. My repo doesn't do all this yet.

**Comment by davenza** [comment]
@timkpaine I thought about a similar design, but I don't think it will work.

You link your python library dynamically against libarrow and libarrow_python. But, you link against _specific versions_ of libarrow and libarrow_python. If the pyarrow version changes, pyarrow will load a different libarrow.so and libarrow_python.so, and the Python library will crash because there is no ABI stability. I have tested this myself in a little project

Maybe there is a compilation/linker flag I'm not aware of that can solve this. I am very interested to see how your minimal work example could be completed.

**Comment by timkpaine** [comment]
ref: https://github.com/apache/arrow/pull/37797

**Comment by davenza** [comment]
@timkpaine thank you! I think this solves my problem. I hope this gets merged soon.

**Comment by timkpaine** [comment]
@davenza https://github.com/timkpaine/arrow-cpp-python-nocopy/pull/3 removes the need to have pyarrow, and should solve all ABI problems as the linkage between the python layer and the C++ layer is completely separated by the C API layer. I'm only doing a raw CPython binding for now, will do pybind after I finish array and table support.

**Comment by timkpaine** [comment]
ok pybind is done too, this is pretty easy and seems correct.

@pitrou sorry to bother you, does this look like the current best practice? https://github.com/timkpaine/arrow-cpp-python-nocopy/pull/3

**Comment by pitrou** [comment]
@timkpaine I see three problems:
1) pack/unpack will not roundtrip, see https://github.com/apache/arrow/issues/10488#issuecomment-1964585311
2) that code seems to lack error detection and handling, and might have reference leaks
3) the no-op capsule destructors (`ReleaseArrowArrayPyCapsule` and `ReleaseArrowSchemaPyCapsule`) will lead to massive memory leaks. Why didn't you reuse the code in https://arrow.apache.org/docs/format/CDataInterface/PyCapsuleInterface.html#create-a-pycapsule?

**Comment by timkpaine** [comment]
Thanks for looking, sounds like it's good in general, this is just a demo so I just wanted to demonstrate working.

1. I will adjust to check if it has those OR if it's already a correct capsule
2. This I'll add later
3. Laziness during debugging, was getting segfaults but it ended up being that c array returns a tuple and I didn't read close enough at the type annotations.

**Comment by dhirschfeld** [comment]
From reading this issue, my understanding is that, at present, the only way to pass a `pyarrow.Table` to `perspective` is to first convert it to a binary buffer as shown in the original post - is that correct?

**Comment by mhkeller** [comment]
For anyone looking for how to do this in Node.js, this seems to work

```js
const table = new arrow.Table(recordBatches);
const ipcStream = arrow.tableToIPC(table, 'stream');
const bytes = Buffer.from(ipcStream, 'utf-8');
```

If there's a better way in Node to pass the table directly, please let me know!

**Comment by texodus** [comment]
I've updated `perspective-python` docs and tests to reflect that this now works (and has since ~`v3.0.0`) in #3172

This implementation does _not_ transfer the pointer unfortunately. I'm not quite sure with modern Arrow whether or not the problem description for this even still applies, but I think its worth separating the issue at least. So, I'm going to close this issue to reflect the API contract is now fulfilled, and further discussion of an internal implementation for performance should go to a new Issue.

@mhkeller This will never work in JavaScript with our static bundles - WASM heaps cannot be shared.
